ResetPasswordBundle/src/PasswordResetHelper.php

```diff
@@ -76,67 +109,67 @@ public function generateResetToken(object $user): PasswordResetToken
         );
     }
 
-    public function validateTokenAndFetchUser(string $fullToken): object
+    /**
+     * Validate a PasswordResetRequest and fetch user from database
+     *
+     * @param string $fullToken selector + non-hashed verifier token
+     * @return UserInterface
+     * @throws ExpiredResetPasswordTokenException
+     * @throws InvalidResetPasswordTokenException
+     * @throws \Throwable @TODO Refactor TokenGenerator and change exception signature
+     */
+    public function validateTokenAndFetchUser(string $fullToken): UserInterface
     {
-        $resetToken = $this->findToken($fullToken);
+        /** @var PasswordResetRequestInterface $resetToken */
+        $resetRequest = $this->findToken($fullToken);
 
-        if ($resetToken->isExpired()) {
+        if ($resetRequest->isExpired()) {
             throw new ExpiredResetPasswordTokenException();
         }
 
+        /** @var UserInterface $user */
+        $user = $resetRequest->getUser();
+
         $verifierToken = substr($fullToken, self::SELECTOR_LENGTH);
-        $hashedVerifierToken = $this->hashVerifierToken(
+
+        $hashedVerifierToken = $this->tokenGenerator->getToken(
+            $this->tokenSigningKey,
+            $resetRequest->getExpiresAt(),
             $verifierToken,
-            $this->repository->getUserIdentifier($resetToken->getUser()),
-            $resetToken->getExpiresAt()
+            $user->getId()
         );
 
-        if (false === hash_equals($resetToken->getToken(), $hashedVerifierToken)) {
+        if (false === hash_equals($resetRequest->getHashedToken(), $hashedVerifierToken)) {
             throw new InvalidResetPasswordTokenException();
         }
 
-        return $resetToken->getUser();
-    }
-
-    private function findToken(string $token): PasswordResetRequestInterface
-    {
-        $selector = substr($token, 0, self::SELECTOR_LENGTH);
-
-        return $this->repository->findPasswordResetRequest($selector);
+        return $user;
     }
 
     /**
-     * Original credit to Laravel's Str::random() method.
+     * Remove a single PasswordResetRequest object from the database
+     *
+     * @param string $fullToken selector + non-hashed verifier token
+     * @throws InvalidResetPasswordTokenException
      */
-    private function generateRandomAlphaNumericString(int $length)
+    public function removeResetRequest(string $fullToken): void
```

Comment:
I think this method isn't used yet? And it might not be needed (and not on the interface either). We DO need to call `$this->repository->removeResetRequest($request);` after a token is used - probably in `validateTokenAndFetchUser` - so that it can't be used again. Hmm, and maybe rename that method to `consumeTokenAndFetchUser`  - it makes it clear that the token will not be useable after calling this method.

Reply:
done - to be used in the app side.. safer than calling `$this->repository->removeResetRequest($request);` as previously discussed.. left method name the same for now, as we are not destroying the request within `validateTokenAndFetchUser()`